### PR TITLE
[Coverage] ApplicationEngine Runtime/Storage/Contract behavioral tests

### LINKED ISSUE #4693: Review tracker: small independent unit-test coverage PRs (toward ~90% line coverage)
## Summary

First, **I'm sorry for the large number of open PRs.**

I've been investing in **Neo unit-test coverage for about six months**, with the goal of moving line coverage from roughly **~83% toward ~90%**. Recently I opened a large set of **small, independent, test-only** PRs so each change is easy to review and can merge without stacking or rework.

This issue is a **review index** for that batch: please treat the volume as intentional (many tiny slices), not as a demand to merge everything at once.

---

## Why so many small PRs?

| Design choice | Why |
|---|---|
| **One exclusive test file per PR** | Reviews stay small; conflicts stay rare |
| **Each branch based on `master-n3`** | PRs do **not** depend on each other |
| **Tests only under `tests/**`** | No product behavior changes in coverage PRs |
| **Bugfixes separate** | If a test exposed a real bug, product fixes went to separate PRs |

**Scale of this batch**

- **97 open coverage PRs** (`[Coverage 1/N]` … `[Coverage 97/N]`)
- **~6,100 lines** of new test code (sum of PR additions; ~**60 lines average** per PR)
- Almost every PR adds a single new `UT_*.cs` file

---

## Estimated coverage impact (after all of these merge)

| | Estimate |
|---|---|
| **Recent baseline** (coverlet, Neo monorepo / `Neo` assembly heavy) | **~83%** line coverage |
| **Expected after this 97-PR batch alone** | **~84–86%** line coverage (modest lift) |
| **Campaign target** | **~90%** line coverage |

**Important honesty:** this batch mostly covers **pure helpers, enums, serializers, builders, caches, and small payload types**. That is useful, but it will **not** by itself close the gap to ~90%. The remaining headroom is dominated by larger areas (`ApplicationEngine`, native contracts, P2P protocol, ledger/mempool).

So: please review/merge these when convenient for quality and safety; they are **incremental**, not a claim that coverage work is finished.

---

## Suggested review priority

If you only have time for a subset, please start with **higher-signal** PRs (more behavioral coverage, not just enum constants):

### Priority A — higher value (behavior / engine / storage / ABI)

| PR | Focus |
|---|---|
| [#4680](https://github.com/neo-project/neo/pull/4680) | ApplicationEngine fault helpers (`GetEngineStackInfoOnFault` / exception info) |
| [#4691](https://github.com/neo-project/neo/pull/4691) | `ApplicationExecuted` from engine state |
| [#4636](https://github.com/neo-project/neo/pull/4636) | `ContractAbi` GetMethod / JSON / StackItem |
| [#4642](https://github.com/neo-project/neo/pull/4642) | `ContractParameter.ToStackItem` |
| [#4643](https://github.com/neo-project/neo/pull/4643) | `InteropParameterDescriptor` converters |
| [#4663](https://github.com/neo-project/neo/pull/4663) | `InteropDescriptor` hash / parameters |
| [#4664](https://github.com/neo-project/neo/pull/4664) | `ExecutionContextState` |
| [#4671](https://github.com/neo-project/neo/pull/4671) | `StoreCache` read/find |
| [#4678](https://github.com/neo-project/neo/pull/4678) | `NeoSystemExtensions` (time / valid-until / traceable) |
| [#4617](https://github.com/neo-project/neo/pull/4617) | `BinarySerializer` deserialize round-trip |
| [#4616](https://github.com/neo-project/neo/pull/4616) | `Contract.Create` by hash / multisig validation |
| [#4596](https://github.com/neo-project/neo/pull/4596) | Witness condition serialize / JSON / StackItem |
| [#4690](https://github.com/neo-project/neo/pull/4690) | ScriptBuilder `EmitPush` / `EmitSysCall` |
| [#4684](https://github.com/neo-project/neo/pull/4684) | `EvaluationStack.ToJson` |
| [#4679](https://github.com/neo-project/neo/pull/4679) | `PriorityMessageQueue` |
| [#4652](https://github.com/neo-project/neo/pull/4652) | `MemoryExtensions` AsSerializable |
| [#4655](https://github.com/neo-project/neo/pull/4655) | `BinaryReaderExtensions` |
| [#4672](https://github.com/neo-project/neo/pull/4672) | `BinaryWriterExtensions` |
| [#4667](https://github.com/neo-project/neo/pull/4667) | `MemoryReaderExtensions` |
| [#4692](https://github.com/neo-project/neo/pull/4692) | `SpanExtensions` LZ4 edges |

### Priority B — solid mid-tier (payloads, wallet, crypto helpers)

| PR | Focus |
|---|---|
| [#4634](https://github.com/neo-project/neo/pull/4634) | `PingPayload` |
| [#4601](https://github.com/neo-project/neo/pull/4601) | `OracleResponse` serialize / JSON |
| [#4597](https://github.com/neo-project/neo/pull/4597) | `OracleRequest` stack item |
| [#4658](https://github.com/neo-project/neo/pull/4658) | `Conflicts` ToJson / serialize |
| [#4627](https://github.com/neo-project/neo/pull/4627) | `ContractMethodDescriptor` |
| [#4625](https://github.com/neo-project/neo/pull/4625) | `ContractGroup` FromJson |
| [#4630](https://github.com/neo-project/neo/pull/4630) | `ContractPermissionDescriptor` edges |
| [#4638](https://github.com/neo-project/neo/pull/4638) | `WhitelistedContract` stack item |
| [#4669](https://github.com/neo-project/neo/pull/4669) | `NEP6WalletFactory` |
| [#4675](https://github.com/neo-project/neo/pull/4675) | `Wallet.Create` / `Open` |
| [#4660](https://github.com/neo-project/neo/pull/4660) | BIP32 `KeyPath` |
| [#4653](https://github.com/neo-project/neo/pull/4653) | `ECCurve` |
| [#4595](https://github.com/neo-project/neo/pull/4595) | `ECFieldElement` expansions |
| [#4649](https://github.com/neo-project/neo/pull/4649) | `SignerManager` |
| [#4659](https://github.com/neo-project/neo/pull/4659) | `StoreFactory` |
| [#4687](https://github.com/neo-project/neo/pull/4687) | `FIFOCache` |
| [#4668](https://github.com/neo-project/neo/pull/4668) | `ECPointCache` capacity |
| [#4666](https://github.com/neo-project/neo/pull/4666) | `ECDsaCache` |

### Priority C — quick enum / constant / property slices (fastest reviews)

These are intentionally tiny (often wire-value asserts). Safe to batch-review:

| PR | Focus |
|---|---|
| [#4608](https://github.com/neo-project/neo/pull/4608) | Smart contract flags / related enums |
| [#4613](https://github.com/neo-project/neo/pull/4613) | Ledger-related enums |
| [#4619](https://github.com/neo-project/neo/pull/4619) | P2P / plugin enums |
| [#4620](https://github.com/neo-project/neo/pull/4620) | `ContractParameterType` |
| [#4624](https://github.com/neo-project/neo/pull/4624) | Inventory / attribute / capability type enums |
| [#4629](https://github.com/neo-project/neo/pull/4629) | `OracleResponseCode` |
| [#4635](https://github.com/neo-project/neo/pull/4635) | `WitnessConditionType` |
| [#4637](https://github.com/neo-project/neo/pull/4637) | `TrackState` |
| [#4639](https://github.com/neo-project/neo/pull/4639) | `SeekDirection` |
| [#4640](https://github.com/neo-project/neo/pull/4640) | `HashAlgorithm` |
| [#4645](https://github.com/neo-project/neo/pull/4645) | `LogLevel` |
| [#4654](https://github.com/neo-project/neo/pull/4654) | `Hasher` |
| [#4657](https://github.com/neo-project/neo/pull/4657) | full `MessageCommand` map |
| [#4661](https://github.com/neo-project/neo/pull/4661) | `MessageFlags` |
| [#4689](https://github.com/neo-project/neo/pull/4689) | `JPathTokenType` |
|| …plus remaining enum/default slices in the full table below |

---

## Related bugfix PRs (not coverage)

If tests uncovered real product issues, fixes were opened **separately** (please review these as normal bugfixes, not coverage noise):

- [#4598](https://github.com/neo-project/neo/pull/4598) — GroupCondition equality (`ECPoint.Equals`)
- [#4618](https://github.com/neo-project/neo/pull/4618) — `ContractEventDescriptor.GetHashCode` (parameter content)

---

## Full index of open coverage PRs

| PR | Focus |
|---|---|
| [#4593](https://github.com/neo-project/neo/pull/4593) | Add more StringExtensions unit tests |
| [#4594](https://github.com/neo-project/neo/pull/4594) | Add AndConditionBuilder and OrConditionBuilder tests |
| [#4595](https://github.com/neo-project/neo/pull/4595) | Expand ECFieldElement unit tests |
| [#4596](https://github.com/neo-project/neo/pull/4596) | Add serialize/JSON/StackItem tests for witness conditions |
| [#4597](https://github.com/neo-project/neo/pull/4597) | Add OracleRequest stack item round-trip tests |
| [#4599](https://github.com/neo-project/neo/pull/4599) | Add wallet ToAddress and XOR unit tests |
| [#4600](https://github.com/neo-project/neo/pull/4600) | Add ContractParameterDefinition unit tests |
| [#4601](https://github.com/neo-project/neo/pull/4601) | Add OracleResponse serialize and JSON tests |
| [#4602](https://github.com/neo-project/neo/pull/4602) | Add WitnessRuleBuilder edge case tests |
| [#4603](https://github.com/neo-project/neo/pull/4603) | Add ChannelsConfig defaults and MessageFlags tests |
| [#4604](https://github.com/neo-project/neo/pull/4604) | Add JsonSerializer.SerializeToByteArray tests |
| [#4605](https://github.com/neo-project/neo/pull/4605) | Add KeyBuilder edge case tests |
| [#4606](https://github.com/neo-project/neo/pull/4606) | Add MethodToken ToJson and underscore method tests |
| [#4607](https://github.com/neo-project/neo/pull/4607) | Add NotifyEventArgs StackItem conversion tests |
| [#4608](https://github.com/neo-project/neo/pull/4608) | Add tests for smart contract flags and related enums |
| [#4609](https://github.com/neo-project/neo/pull/4609) | Add ContractBasicMethod constant tests |
| [#4610](https://github.com/neo-project/neo/pull/4610) | Add StorageContext unit tests |
| [#4611](https://github.com/neo-project/neo/pull/4611) | Add TimeProvider unit tests |
| [#4612](https://github.com/neo-project/neo/pull/4612) | Add MaxLengthAttribute unit tests |
| [#4613](https://github.com/neo-project/neo/pull/4613) | Add ledger and related enum tests |
| [#4614](https://github.com/neo-project/neo/pull/4614) | Add AccountState stack item tests |
| [#4615](https://github.com/neo-project/neo/pull/4615) | Add SmartContract Helper fee cost tests |
| [#4616](https://github.com/neo-project/neo/pull/4616) | Add Contract.Create by hash and multisig validation tests |
| [#4617](https://github.com/neo-project/neo/pull/4617) | Add BinarySerializer deserialize round-trip tests |
| [#4619](https://github.com/neo-project/neo/pull/4619) | Add P2P and plugin enum tests |
| [#4620](https://github.com/neo-project/neo/pull/4620) | Add ContractParameterType value tests |
| [#4621](https://github.com/neo-project/neo/pull/4621) | Add SignException unit tests |
| [#4622](https://github.com/neo-project/neo/pull/4622) | Add ContractEventDescriptor edge tests |
| [#4623](https://github.com/neo-project/neo/pull/4623) | Add RelayCache capacity eviction tests |
| [#4624](https://github.com/neo-project/neo/pull/4624) | Add InventoryType, TransactionAttributeType, NodeCapabilityType tests |
| [#4625](https://github.com/neo-project/neo/pull/4625) | Add ContractGroup FromJson tests |
| [#4626](https://github.com/neo-project/neo/pull/4626) | Add StorageItem sealed interoperable tests |
| [#4627](https://github.com/neo-project/neo/pull/4627) | Add ContractMethodDescriptor unit tests |
| [#4628](https://github.com/neo-project/neo/pull/4628) | Add BigDecimal edge case tests |
| [#4629](https://github.com/neo-project/neo/pull/4629) | Add OracleResponseCode value tests |
| [#4630](https://github.com/neo-project/neo/pull/4630) | Add ContractPermissionDescriptor edge tests |
| [#4631](https://github.com/neo-project/neo/pull/4631) | Add WildcardContainer ToJson tests |
| [#4632](https://github.com/neo-project/neo/pull/4632) | Add BigDecimal Parse/TryParse tests |
| [#4633](https://github.com/neo-project/neo/pull/4633) | Add BloomFilter post-Add bit tests |
| [#4634](https://github.com/neo-project/neo/pull/4634) | Add PingPayload unit tests |
| [#4635](https://github.com/neo-project/neo/pull/4635) | Add WitnessConditionType value tests |
| [#4636](https://github.com/neo-project/neo/pull/4636) | Add ContractAbi unit tests |
| [#4637](https://github.com/neo-project/neo/pull/4637) | Add TrackState value tests |
| [#4638](https://github.com/neo-project/neo/pull/4638) | Add WhitelistedContract stack item tests |
| [#4639](https://github.com/neo-project/neo/pull/4639) | Add SeekDirection value tests |
| [#4640](https://github.com/neo-project/neo/pull/4640) | Add HashAlgorithm value tests |
| [#4641](https://github.com/neo-project/neo/pull/4641) | Add MemoryStoreProvider unit tests |
| [#4642](https://github.com/neo-project/neo/pull/4642) | Add ContractParameter ToStackItem tests |
| [#4643](https://github.com/neo-project/neo/pull/4643) | Add InteropParameterDescriptor unit tests |
| [#4644](https://github.com/neo-project/neo/pull/4644) | Add UInt160.MakeScript tests |
| [#4645](https://github.com/neo-project/neo/pull/4645) | Add LogLevel value tests |
| [#4646](https://github.com/neo-project/neo/pull/4646) | Add Utility.Log and StrictUTF8 tests |
| [#4647](https://github.com/neo-project/neo/pull/4647) | Add TransferOutput unit tests |
| [#4648](https://github.com/neo-project/neo/pull/4648) | Add ICollection GetVarSize and ToByteArray tests |
| [#4649](https://github.com/neo-project/neo/pull/4649) | Add SignerManager unit tests |
| [#4650](https://github.com/neo-project/neo/pull/4650) | Add TransactionRemovedEventArgs tests |
| [#4651](https://github.com/neo-project/neo/pull/4651) | Add NewTransactionEventArgs tests |
| [#4652](https://github.com/neo-project/neo/pull/4652) | Add MemoryExtensions unit tests |
| [#4653](https://github.com/neo-project/neo/pull/4653) | Add ECCurve unit tests |
| [#4654](https://github.com/neo-project/neo/pull/4654) | Add Hasher enum value tests |
| [#4655](https://github.com/neo-project/neo/pull/4655) | Add BinaryReaderExtensions unit tests |
| [#4656](https://github.com/neo-project/neo/pull/4656) | Add ContractTaskMethodBuilder unit tests |
| [#4657](https://github.com/neo-project/neo/pull/4657) | Add full MessageCommand value tests |
| [#4658](https://github.com/neo-project/neo/pull/4658) | Add Conflicts ToJson and serialize tests |
| [#4659](https://github.com/neo-project/neo/pull/4659) | Add StoreFactory unit tests |
| [#4660](https://github.com/neo-project/neo/pull/4660) | Add BIP32 KeyPath unit tests |
| [#4661](https://github.com/neo-project/neo/pull/4661) | Add MessageFlags value tests |
| [#4662](https://github.com/neo-project/neo/pull/4662) | Add ContractTask unit tests |
| [#4663](https://github.com/neo-project/neo/pull/4663) | Add InteropDescriptor unit tests |
| [#4664](https://github.com/neo-project/neo/pull/4664) | Add ExecutionContextState unit tests |
| [#4666](https://github.com/neo-project/neo/pull/4666) | Add ECDsaCache unit tests |
| [#4667](https://github.com/neo-project/neo/pull/4667) | Add MemoryReaderExtensions unit tests |
| [#4668](https://github.com/neo-project/neo/pull/4668) | Add ECPointCache capacity eviction tests |
| [#4669](https://github.com/neo-project/neo/pull/4669) | Add NEP6WalletFactory unit tests |
| [#4670](https://github.com/neo-project/neo/pull/4670) | Add TaskSession HasTooManyTasks tests |
| [#4671](https://github.com/neo-project/neo/pull/4671) | Add StoreCache unit tests |
| [#4672](https://github.com/neo-project/neo/pull/4672) | Add BinaryWriterExtensions unit tests |
| [#4673](https://github.com/neo-project/neo/pull/4673) | Add IInteroperable Clone/FromReplica tests |
| [#4674](https://github.com/neo-project/neo/pull/4674) | Add Idle actor message tests |
| [#4675](https://github.com/neo-project/neo/pull/4675) | Add Wallet.Create and Open tests |
| [#4676](https://github.com/neo-project/neo/pull/4676) | Add DisableCompressionCapability unit tests |
| [#4677](https://github.com/neo-project/neo/pull/4677) | Add ContractTaskAwaiter OnCompleted tests |
| [#4678](https://github.com/neo-project/neo/pull/4678) | Add NeoSystemExtensions unit tests |
| [#4679](https://github.com/neo-project/neo/pull/4679) | Add PriorityMessageQueue unit tests |
| [#4680](https://github.com/neo-project/neo/pull/4680) | Add ApplicationEngine fault helper tests |
| [#4681](https://github.com/neo-project/neo/pull/4681) | Add TransactionRouter record tests |
| [#4682](https://github.com/neo-project/neo/pull/4682) | Add Treasury native contract tests |
| [#4683](https://github.com/neo-project/neo/pull/4683) | Add UnknownCapability edge tests |
| [#4684](https://github.com/neo-project/neo/pull/4684) | Add EvaluationStack.ToJson tests |
| [#4685](https://github.com/neo-project/neo/pull/4685) | Add ScriptBuilderExtensions edge tests |
| [#4686](https://github.com/neo-project/neo/pull/4686) | Add PriorityMailbox default tests |
| [#4687](https://github.com/neo-project/neo/pull/4687) | Add FIFOCache unit tests |
| [#4688](https://github.com/neo-project/neo/pull/4688) | Add ReflectionCacheAttribute unit tests |
| [#4689](https://github.com/neo-project/neo/pull/4689) | Add JPathTokenType value tests |
| [#4690](https://github.com/neo-project/neo/pull/4690) | Add ScriptBuilder EmitPush parameter tests |
| [#4691](https://github.com/neo-project/neo/pull/4691) | Add ApplicationExecuted unit tests |
| [#4692](https://github.com/neo-project/neo/pull/4692) | Add SpanExtensions LZ4 edge tests |

---

## How to review efficiently

1. Prefer **Priority A** first if time is limited.  
2. Priority C can often be reviewed in bulk (pattern is repetitive by design).  
3. Each PR is independent — merge order does **not** matter.  
4. Reject / request changes on any individual PR without blocking the rest.  

Again: **sorry for the volume**, and thank you for any review bandwidth you can spare. The intent is higher confidence in Neo's core helpers while keeping each review bite-sized.

/cc maintainers reviewing `master-n3` test coverage

### PR DESCRIPTION
**Pull request by cschuchardt88**
# Description

Adds exclusive ApplicationEngine behavioral coverage for high-ROI Runtime, Storage, and Contract interop paths, plus a small shared `TestEngineRunner` fixture. Part of the larger ~90% coverage track. Test-only; no product behavior changes.

Related: #4693

# Change Log

- Add File `tests/Neo.UnitTests/SmartContract/TestEngineRunner.cs`
- Add File `tests/Neo.UnitTests/SmartContract/UT_ApplicationEngine_Runtime_Coverage.cs`
- Add File `tests/Neo.UnitTests/SmartContract/UT_ApplicationEngine_Storage_Coverage.cs`
- Add File `tests/Neo.UnitTests/SmartContract/UT_ApplicationEngine_Contract_Coverage.cs`

Fixes #4693 (partial — ApplicationEngine slice)

## Type of change

- [ ] Optimization (the change is only an optimization)
- [ ] Style (the change is only a code style for better maintenance or standard purpose)
- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] This change requires a documentation update

# How Has This Been Tested?

- [x] Unit Testing
- [ ] Run Application
- [ ] Local Computer Tests
- [ ] No Testing

`dotnet test tests/Neo.UnitTests --filter FullyQualifiedName~UT_ApplicationEngine_`

# Checklist:

- [x] My code follows the style guidelines of this project
- [x] I have performed a self-review of my code
- [ ] I have commented my code, particularly in hard-to-understand areas
- [ ] I have made corresponding changes to the documentation
- [x] My changes generate no new warnings
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] New and existing unit tests pass locally with my changes
- [ ] Any dependent changes have been merged and published in downstream modules

### CODE REVIEW

File: tests/Neo.UnitTests/SmartContract/UT_ApplicationEngine_Contract_Coverage.cs [hunk 1]
```diff
@@ -0,0 +1,107 @@
...
+        public void Setup()
+        {
+            _snapshot = TestBlockchain.GetTestSnapshotCache().CloneCache();
+        }
+
+        [TestMethod]
+        public void CallContract_UnderscoreMethod_Throws()
+        {
+            using var engine = TestEngineRunner.CreateWithScript(_snapshot, new byte[] { (byte)OpCode.NOP });
+            Assert.ThrowsExactly<ArgumentException>(() =>
```
Comment (shargon) [inline tests/Neo.UnitTests/SmartContract/UT_ApplicationEngine_Contract_Coverage.cs:39]: You should try to call a valid underscore method, otherwise the ut can fail because the method doesn't exists

Comment (cschuchardt88) [reply tests/Neo.UnitTests/SmartContract/UT_ApplicationEngine_Contract_Coverage.cs:39]: Partly on intent, not on current behavior.

`CallContract` checks `method.StartsWith('_')` and throws `ArgumentException` **before** contract/method lookup, so `"_hidden"` cannot fail as "method doesn't exist" with today's control flow — the underscore guard always wins and the UT is stable against the real path.

What you are asking for is stronger test intent: protocol-reserved names like `_initialize` / `_deploy` make it obvious we are locking the ban on calling those via `System.Contract.Call`, not a made-up name that would also fail lookup if the check were reordered or removed.

So: not a flake bug today; a clarity / future-proofing nit. Happy to switch the call to `ContractBasicMethod.Initialize` (or `Deploy`) if you want that clearer naming.

File: tests/Neo.UnitTests/SmartContract/UT_ApplicationEngine_Runtime_Coverage.cs [hunk 2]
```diff
@@ -0,0 +1,142 @@
...
+
+        [TestMethod]
+        public void GetRandom_ReturnsNonNegative()
+        {
+            using var engine = TestEngineRunner.CreateWithScript(_snapshot, new byte[] { (byte)OpCode.NOP }, gas: 100_0000_0000);
+            var a = engine.GetRandom();
+            var b = engine.GetRandom();
+            Assert.IsTrue(a >= 0);
+            Assert.IsTrue(b >= 0);
+        }
```
Comment (ajara87) [inline tests/Neo.UnitTests/SmartContract/UT_ApplicationEngine_Runtime_Coverage.cs:114]: Since GetRandom() returns an unsigned BigInteger the non-negative assertions are guaranteed. Could we instead verify that each call increases FeeConsumed to test an actual observable behavior?

**Review by ajara87 — DISMISSED** [review]

File: tests/Neo.UnitTests/SmartContract/UT_ApplicationEngine_Runtime_Coverage.cs (hunk 2 above)
Comment (cschuchardt88) [reply tests/Neo.UnitTests/SmartContract/UT_ApplicationEngine_Runtime_Coverage.cs:114]: Agreed — `GetRandom` builds an unsigned `BigInteger`, so non-negativity is not a useful assert.

**Fixed**: renamed to `GetRandom_IncreasesFeeConsumed` and now checks that `FeeConsumed` rises after each call (the real observable side effect via `AddFee`).

**Review by ajara87 — APPROVED** [review]

**Review by shargon — APPROVED** [review]
